pkg/handler/handler.go

```diff
@@ -55,13 +55,24 @@ type OpenAPIService struct {
 
 	lastModified time.Time
 
-	specBytes []byte
-	specPb    []byte
-	specPbGz  []byte
+	jsonCache    cache
+	protoCache   cache
+	protoGzCache cache
+}
+
+type cache struct {
+	BuildCache func() ([]byte, string, error)
+	once       sync.Once
+	bytes      []byte
+	etag       string
+	err        error
+}
 
-	specBytesETag string
-	specPbETag    string
-	specPbGzETag  string
+func (c *cache) Get() ([]byte, string, error) {
+	c.once.Do(func() {
+		c.bytes, c.etag, c.err = c.BuildCache()
```

Comment:
Maybe something like this:
```suggestion
		c.bytes, c.err = c.BuildCache()
		if c.err != nil {
		    c.etag = computeETag(c.bytes)
		}
```